bus-mapping/src/circuit_input_builder.rs

```diff
@@ -244,3 +317,1048 @@ impl<'a> CircuitInputBuilder {
+        } else if error.starts_with(GETH_ERR_STACK_UNDERFLOW) {
+            ExecError::StackUnderflow
+        } else if error.starts_with(GETH_ERR_INVALID_OPCODE) {
+            ExecError::InvalidOpcode
+        } else {
+            panic!("Unknown GethExecStep.error: {}", error);
+        }
+    })
+}
+
+impl<'a> CircuitInputStateRef<'a> {
+    fn get_step_err(
+        &self,
+        step: &GethExecStep,
+        next_step: Option<&GethExecStep>,
+    ) -> Option<ExecError> {
+        if let Some(error) = get_step_reported_error(step) {
+            return Some(error);
+        }
+        assert!(step.error.is_none());
+
+        let next_depth = next_step.map(|s| s.depth).unwrap_or(0);
+        let next_result = next_step
+            .map(|s| s.stack.last().unwrap_or_else(|_| Word::zero()))
+            .unwrap_or_else(Word::zero);
+
+        // Return from a call with a failure, without calling RETURN
+        if step.op != OpcodeId::RETURN
+            && step.depth != next_depth
+            && next_result == Word::zero()
+        {
+            return Some(match step.op {
+                OpcodeId::REVERT => ExecError::ExecutionReverted,
+                OpcodeId::JUMP | OpcodeId::JUMPI => ExecError::InvalidJump,
+                OpcodeId::RETURNDATACOPY => ExecError::ReturnDataOutOfBounds,
+                _ => {
+                    panic!("Cannot figure out call failure error in {:?}", step)
+                }
+            });
+        }
+
+        // Return from a {CREATE, CREATE2} with a failure, via RETURN
+        // TODO: if not self.tx_ctx.is_root and self.tx_ctx.is_create {
+        if step.op == OpcodeId::RETURN && next_result == Word::zero() {
+            let offset = step.stack.nth_last(0).unwrap();
+            let length = step.stack.nth_last(1).unwrap();
+            if length > Word::from(0x6000) {
+                return Some(ExecError::MaxCodeSizeExceeded);
+            } else if length > Word::zero()
+                && !step.memory.0.is_empty()
+                && step.memory.0.get(offset.low_u64() as usize) == Some(&0xef)
+            {
+                return Some(ExecError::InvalidCode);
+            } else if Word::from(200) * length > Word::from(step.gas.0) {
+                return Some(ExecError::CodeStoreOutOfGas);
+            } else {
+                panic!("Cannot figure out RETURN error in {:?}", step);
+            }
+        }
+        // }
+
+        // The *CALL* code was not executed
+        let next_pc = next_step.map(|s| s.pc.0).unwrap_or(1);
+        if [
+            OpcodeId::CALL,
+            OpcodeId::CALLCODE,
+            OpcodeId::DELEGATECALL,
+            OpcodeId::STATICCALL,
+            OpcodeId::CREATE,
+            OpcodeId::CREATE2,
+        ]
+        .contains(&step.op)
+            && next_result == Word::zero()
+            && next_pc != 0
+        {
+            if step.depth == 1025 {
+                return Some(ExecError::Depth);
+            }
```

Comment:
Shouldn't we check `> 1024` instead. What would happen if we get `1026` here?

Reply:
There's no need for that, because geth will never reach a depth > 1025.  When depth = 1025, any function that would increment the depth will fail.

Also, note that it's 1025 because geth sets depth = 1 to the first step, so after the first step, the depth can be increased by 1024.